File: Lab 11/t1.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/**
 * A point.
 */
struct _Point {
  int x; //< X axis.
  int y; //< Y Axis.
};
typedef struct _Point Point;

/**
 * A line, consisting of 2 points.
 *
 * @see struct _Point
 */
struct _Line {
  struct _Point p1; //< Point 1.
  struct _Point p2; //< Point 2.
};
typedef struct _Line Line;

/**
 * A triangle, consisting of 3 points.
 *
 * @see struct _Point
 */
struct _Triangle {
  struct _Point p1; //< Point 1.
  struct _Point p2; //< Point 2.
  struct _Point p3; //< Point 3.
};
typedef struct _Triangle Triangle;

/**
 * Given a line, calculate its length.
 *
 * @param line [in] A line.
 * @return Length of line, in double.
 * @see Line
 */
static double length (Line line);

/**
 * Given a triangle, calculate its circumference.
 *
 * @param triangle [in] A triangle.
 * @return Circumference of triangle, in double.
 * @see Triangle
 */
static double circumference (Triangle triangle);

/**
 * Print point information to stdout.
 *
 * The format is: "Point: (x, y)".
 *
 * @param point [in] Point to be printed.
 */
static void printPoint (Point point);

/**
 * Print line information to stdout.
 *
 * The format is: "Line: (x1, y1) - (x2, y2) Length: len".
 *
 * @param line [in] Line to be printed.
 */
static void printLine (Line line);

/**
 * Print triangle information to stdout.
 *
 * The format is: "Triangle: (x1, y1) - (x2, y2) - (x3, y3) Circumference: c".
 *
 * @param triangle [in] Triangle to be printed.
 */
static void printTriangle (Triangle triangle);

int main (void)
{
  // Per requirement.
  Point p[6] = {
    {1, 5}, {2, 1}, {2, 7}, {4, 3}, {6, 8}, {8, 2}
  };
  Line l[4] = {
    {p[0], p[2]}, {p[3], p[4]}, {p[3], p[5]}, {p[4], p[5]}
  };
  Triangle t = {p[3], p[4], p[5]};

  for (int i = 0; i < 6; i++) {
    printPoint (p[i]);
  }
  for (int i = 0; i < 4; i++) {
    printLine (l[i]);
  }
  printTriangle (t);

  return 0;
}

static double length (Line line)
{
  int delta_x = abs (line.p1.x - line.p2.x);
  int delta_y = abs (line.p1.y - line.p2.y);

  return sqrt (delta_x * delta_x + delta_y * delta_y);
}

static double circumference (Triangle triangle)
{
  Line lines[3] = {
    {triangle.p1, triangle.p2}, {triangle.p2, triangle.p3}, {triangle.p3, triangle.p1}
  };

  return length (lines[0]) + length (lines[1]) + length (lines[2]);
}

static void printPoint (Point point)
{
  printf ("Point: (%d, %d)\n", point.x, point.y);
}

static void printLine (Line line)
{
  printf (
    "Line: (%d, %d) - (%d, %d) Length = %lf\n",
    line.p1.x, line.p1.y, line.p2.x, line.p2.y, length (line)
  );
}

static void printTriangle (Triangle triangle)
{
  printf (
    "Triangle: (%d, %d) - (%d, %d) - (%d, %d) Circumference: %lf\n",
    triangle.p1.x, triangle.p1.y,
    triangle.p2.x, triangle.p2.y,
    triangle.p3.x, triangle.p3.y,
    circumference (triangle)
  );
}

File: Lab 11/t1.go
package main

import (
  "fmt"
  "math"
)

type point struct {
  x, y int
}

type line struct {
  p1, p2 point
}

type triangle struct {
  p1, p2, p3 point
}

/* Methods for point */

// func (p point) print() {
//   fmt.Printf("Point: (%d, %d)\n", p.x, p.y)
// }

/**
 * "Formatter is the interface implemented by values with a custom formatter.
 * The implementation of Format may call Sprint(f) or Fprint(f) etc. to generate
 * its output."
 *   -- https://golang.org/pkg/fmt/#Formatter
 */
func (p point) Format(f fmt.State, c rune) {
  // NOTE: Theoretically I should implement different style of formatters but
  // here I just cannot be bordered.
  if _, e := f.Write([]byte(fmt.Sprintf("Point: (%d, %d)", p.x, p.y))); e != nil {
    f.Write([]byte(fmt.Sprintf("!%%%cPoint!", c)))
  }
}

/* Methods for line */

func (l line) length() float64 {
  // NOTE: point has member x, y as int
  delta_x := math.Abs(float64(l.p1.x - l.p2.x))
  delta_y := math.Abs(float64(l.p1.y - l.p2.y))

  return math.Sqrt(math.Pow(delta_x, 2) + math.Pow(delta_y, 2))
}

// func (l line) print() {
//   fmt.Printf("Line: (%d, %d) - (%d, %d) Length = %f\n",
//     l.p1.x, l.p1.y, l.p2.x, l.p2.y,
//     l.length())
// }

func (l line) Format(f fmt.State, c rune) {
  output := []byte(fmt.Sprintf("Line: (%d, %d) - (%d, %d) Length = %f",
    l.p1.x, l.p1.y, l.p2.x, l.p2.y, l.length()))

  if _, e := f.Write(output); e != nil {
    f.Write([]byte(fmt.Sprintf("!%%%cLine!", c)))
  }
}

/* Methods for triangle */

func (t triangle) circumference() float64 {
  lines := [3]line{line{t.p1, t.p2}, line{t.p2, t.p3}, line{t.p3, t.p1}}

  return lines[0].length() + lines[1].length() + lines[2].length()
}

// func (t triangle) print() {
//   fmt.Printf("Triangle: (%d, %d) - (%d, %d) - (%d, %d) Circumference: %f\n",
//     t.p1.x, t.p1.y,
//     t.p2.x, t.p2.y,
//     t.p3.x, t.p3.y,
//     t.circumference())
// }

func (t triangle) Format(f fmt.State, c rune) {
  output := []byte(fmt.Sprintf(
    "Triangle: (%d, %d) - (%d, %d) - (%d, %d) Circumference: %f",
    t.p1.x, t.p1.y, t.p2.x, t.p2.y, t.p3.x, t.p3.y, t.circumference()))

  if _, e := f.Write(output); e != nil {
    f.Write([]byte(fmt.Sprintf("!%%%cTriangle!", c)))
  }
}

func main() {
  // Per requirement.
  p := [6]point{
    point{1, 5}, point{2, 1}, point{2, 7},
    point{4, 3}, point{6, 8}, point{8, 2}}
  l := [4]line{
    line{p[0], p[2]}, line{p[3], p[4]},
    line{p[3], p[5]}, line{p[4], p[5]}}
  t := triangle{p[3], p[4], p[5]}

  for _, point := range p {
    // point.print()
    fmt.Println(point)
  }
  for _, line := range l {
    // line.print()
    fmt.Println(line)
  }
  // t.print()
  fmt.Println(t)
}
